pkg/source/source.go

```diff
@@ -370,6 +433,18 @@ func (is *Informer) String() string {
 	return fmt.Sprintf("informer source: %p", is.Informer)
 }
 
+// Stop implements StoppableSource to stop it dynamically.
+func (is *Informer) Stop() error {
+	is.mu.Lock()
+	defer is.mu.Unlock()
+	is.canceled = true
+
+	if is.eventHandlerRegistration != nil {
+		return is.Informer.RemoveEventHandler(is.eventHandlerRegistration)
```

Comment:
None of this actually stops the informer which I'd argue is the more important part - Is that a folllow-up?

Admittedly not easy, we will likely need to do some kind of refcounting

Reply:
Good question. It's hard to determine when should we stop the informer. Not only the informers in cache are not only created by source watches, but also triggered by user's `Get`, `List` calls to DelegatingClient. So we can't manage the lifecycle of informer according to the reference count of how many active watches in it.

Reply:
Thinking out loud here a few options:
1. Ref count for watches + some sort of LRU or timed cache for gets/lists that don't already have informers?
2. Ref count for watches + if a get/list requires a new informer (i.e. a watch didn't already start one), that informer is never removed?

Reply:
Hey! Just found this PR.

Gatekeeper is doing something similar, forking controller-runtime's informer cache to add a function that can remove informers. Would it be possible to have this available before solving the ref count issue? That would at least give users who are comfortable handling that complexity the ability to do so, but probably wouldn't catch unwary users.

Here is the key function we rely on:

https://github.com/open-policy-agent/gatekeeper/blob/8b426fb55da22abc0fe9bc925a3ca1ed08df50fe/third_party/sigs.k8s.io/controller-runtime/pkg/dynamiccache/informer_cache.go#L242-L251

We currently use it by maintaining a separate cache for dynamic watches:

https://github.com/open-policy-agent/gatekeeper/blob/8b426fb55da22abc0fe9bc925a3ca1ed08df50fe/pkg/watch/manager.go#L64-L91

That exports "registrars" to managing controllers to add or remove watches.

https://github.com/open-policy-agent/gatekeeper/blob/8b426fb55da22abc0fe9bc925a3ca1ed08df50fe/pkg/watch/registrar.go#L213-L276

https://github.com/open-policy-agent/gatekeeper/blob/8b426fb55da22abc0fe9bc925a3ca1ed08df50fe/pkg/controller/constrainttemplate/constrainttemplate_controller.go#L570-L582

Static watches (i.e. old-school controller.Watch() and watches initiated by client.Get()) use a separate cache.

Happy to talk more about this model, if interested, or help implementing parts if it means we can stop maintaining a fork!

Reply:
This is great background! I might rebase and continue on the work on this PR (unless @FillZpp has time) to get it to the finish line before 0.15 is released. Feel free to reach out on slack if we want to chat more about it and brainstorm.

Reply:
Happy to brainstorm! (sorry, it's been a busy week, so haven't reached out yet)

Reply:
Sorry for this late reply (I'm on vacation last two weeks). I'll continue on this and get it into 0.15.

About the removal of informer, it seems we could have several ways:

1. Keep ref counts to remove informers automatically, like @joelanford suggested.

> Thinking out loud here a few options:
> 
> 1. Ref count for watches + some sort of LRU or timed cache for gets/lists that don't already have informers?
> 2. Ref count for watches + if a get/list requires a new informer (i.e. a watch didn't already start one), that informer is never removed?

I prefer this way, but I'm not sure will this make users confused? Most of them don't know when and why an informer been removed or not, and the sequence of doing watch and get/list also affect whether the informer will be removed or not...

2. Expose a method to let users manually remove a informer, as @maxsmythe suggested, IIUC.

3. Maybe 1+2 both provided, if they are all necessary and needed.

WDYT @vincepri @alvaroaleman @joelanford @sbueringer

Reply:
In addition, because Gatekeeper uses generic controllers (we have one controller that listens for all constraint kinds), (2) would be a better fit for our model than governing watch livelihood at the controller granularity.

Because dynamic watches are probably best for generic-type controllers (if you have the type hard-coded, why the need for dynamic watches?), managing the watches directly may be a better fit than managing controllers. I think both are workable, but (2) would definitely be less of a reach for us to implement.